operator/src/pod_utils.rs

```diff
@@ -159,115 +157,235 @@ fn create_volume_mounts(log_dir: &Option<String>) -> Vec<VolumeMount> {
-    master_node: &SparkNode,
+    master_urls: &[String],
 ) -> BTreeMap<String, String> {
     let mut labels = BTreeMap::new();
-    labels.insert(TYPE_LABEL.to_string(), node_type.to_string());
-    labels.insert(HASH_LABEL.to_string(), hash.to_string());
-    labels.insert(VERSION_LABEL.to_string(), version.to_string());
+    labels.insert(String::from(APP_COMPONENT_LABEL), node_type.to_string());
+    labels.insert(String::from(APP_ROLE_GROUP_LABEL), role_group.to_string());
+    labels.insert(String::from(APP_INSTANCE_LABEL), cluster_name.to_string());
+
+    labels.insert(APP_VERSION_LABEL.to_string(), version.to_string());
 
     if node_type == &SparkNodeType::Worker {
         labels.insert(
             MASTER_URLS_HASH_LABEL.to_string(),
-            get_hashed_master_urls(master_node),
+            get_hashed_master_urls(master_urls),
         );
     }
 
     labels
 }
 
+/// All pod names follow a simple pattern: <cluster_name>-<role_group>-<node_type>
+///
+/// # Arguments
+/// * `cluster_name` - The name of the cluster as specified in the custom resource
+/// * `role_group` - The role group of the selector
+/// * `node_type` - The cluster node type (e.g. master, worker, history-server)
+///
+pub fn create_pod_name(cluster_name: &str, role_group: &str, node_type: &str) -> String {
+    format!("{}-{}-{}", cluster_name, role_group, node_type).to_lowercase()
+}
+
 /// Get all master urls and hash them. This is required to keep track of which workers
 /// are connected to which masters. In case masters are added / deleted, this hash changes
 /// and we need to restart the worker pods to keep them up to date with all known masters.
 ///
 /// # Arguments
-/// * `master_node` - SparkNode master to retrieve master urls for pod label hash
+/// * `master_urls` - Slice of all known master urls
 ///
-pub fn get_hashed_master_urls(master_node: &SparkNode) -> String {
-    let master_urls = config::get_master_urls(master_node);
+pub fn get_hashed_master_urls(master_urls: &[String]) -> String {
     let mut hasher = DefaultHasher::new();
     for url in master_urls {
         url.hash(&mut hasher);
     }
     hasher.finish().to_string()
 }
 
-/// All pod names follow a simple pattern: <spark_cluster_name>-<node_type>-<selector_hash>-<UUID>
+/// Filter all existing pods for the specified spark node type.
 ///
 /// # Arguments
-/// * `cluster_name` - Current cluster name
-/// * `node_type` - SparkNodeType (master/worker/history-server)
-/// * `hash` - NodeSelector hash
+/// * `pods` - Slice of all existing pods
+/// * `node_type` - The cluster node type (e.g. master, worker, history-server)
 ///
-fn create_pod_name(cluster_name: &str, node_type: &SparkNodeType, hash: &str) -> String {
-    format!(
-        "{}-{}-{}-{}",
-        cluster_name,
-        node_type.as_str(),
-        hash,
-        Uuid::new_v4().as_fields().0.to_string(),
-    )
+pub fn filter_pods_for_type(pods: &[Pod], node_type: &SparkNodeType) -> Vec<Pod> {
```

Comment:
I am not sure about that. I use existing_pods for delete_illegal_pods, wait for terminating and running. So i prefer to keep existing_pods as it is. I could introduce that hashmap, but it will be populated every reconcile, pods cloned etc. 

I think we do not gain much from all that copying only to avoid searching in a (very) small list for some types?

It would be smoother to work with though. I think about it.
@lfrancke you have any opinion on that?

Reply:
I tried to come up with one but I don't really have an opinion, no....I'm fine keeping it as is